### PR TITLE
feat: Make the as prop accept all kind of UrlLike or string

### LINKED ISSUE #517: URL object as a prop
## Is your feature request related to a problem? Please describe.
The problem is related to the error `'as' type must be 'string', but it is object`

## Describe the solution you'd like
Next.js's Link can accept more than a string for the `as` property, it can be a `UrlLike`

For example: `<Link href={...} as={{ pathname: '/test', query: { page: 1 } }}` with output `<a href="/test?page=1">`.

It would be neat to have it in `next-i18next` as well

## Describe alternatives you've considered

The usual `<Link href={...} as={'/test?page=1'}`, but it's a bit more complex to manipulate with dynamic data.

## Additional context
n/a

### PR DESCRIPTION
**Pull request by martpie**
Should fix #517

```tsx
<Link
  href='/somepage'
  as={{ pathname: '/test', hash: 'something', query: { page: 2 } }}
>
  Test
</Link>
```

will output now something like: 

```
http://localhost:3000/en/test?page=2#something
```

### CODE REVIEW

**Comment by martpie** [comment]
I have an additional question: would you accept a PR to enable `strict: true` in TypeScript? I will happily work on making it work, I already saw a couple of potential bugs covered by `strictNullChecks` etc

**Comment by isaachinman** [comment]
> enable strict: true in TypeScript?

Yes.

**Comment by martpie** [comment]
I'll do it in a separate pr then :)

**Comment by martpie** [comment]
Do you need anything else from me @isaachinman ?

File: __tests__/utils/lng-path-corrector.test.ts [hunk 1]
```diff
@@ -35,7 +35,7 @@ describe('lngPathCorrector utility function', () => {
   it('throws if currentRoute.as is not undefined or a string', () => {
```
Comment (isaachinman) [inline __tests__/utils/lng-path-corrector.test.ts]: Needs to be updated.

File: src/utils/lng-path-corrector.ts [hunk 2]
```diff
@@ -1,26 +1,34 @@
...
+}
+
+const parseAs = (originalAs, href): As => {
   const asType = typeof originalAs
   let as
 
   if (asType === 'undefined') {
     as = formatUrl(href, { unicode: true })
-  } else if (asType === 'string') {
+  } else if (asType === 'string' || (asType === 'object' && originalAs !== null)) {
```
Comment (isaachinman) [inline src/utils/lng-path-corrector.ts]: Is checking the type against `object` safe here? Many, many things are `objects` in JavaScript.

File: src/utils/lng-path-corrector.ts [hunk 3]
```diff
@@ -1,26 +1,34 @@
...
   } else {
-    throw new Error(`'as' type must be 'string', but it is ${asType}`)
+    throw new Error(`'as' type must be 'string' or an 'object', but it is ${asType}`)
   }
 
   return as
 }
 
-const parseHref = (originalHref) => {
+const parseHref = (originalHref): UrlObject => {
```
Comment (isaachinman) [inline src/utils/lng-path-corrector.ts]: Is this typing correct? Can't it also be a string?

File: src/utils/lng-path-corrector.ts [hunk 4]
```diff
@@ -58,19 +66,37 @@ export default (config: Config, currentRoute, currentLanguage) => {
     Strip any/all subpaths from the `as` value
   */
   Object.values(localeSubpaths).forEach((subpath) => {
-    if (subpathIsPresent(as, subpath)) {
-      as = removeSubpath(as, subpath)
+    if (typeof as === 'object') {
```
Comment (isaachinman) [inline src/utils/lng-path-corrector.ts]: Shouldn't we also check that `as.pathname` is a string?

File: src/utils/lng-path-corrector.ts [hunk 5]
```diff
@@ -58,19 +66,37 @@ export default (config: Config, currentRoute, currentLanguage) => {
...
 
   if (subpathIsRequired(config, currentLanguage)) {
     const basePath = `${href.protocol}//${href.host}`
-    const currentAs = as.replace(basePath, '')
+    let currentAs = ''
+
     const subpath = subpathFromLng(config, currentLanguage)
 
-    as = `/${subpath}${currentAs}`.replace(/\/$/, '')
+    if (typeof as === 'object') {
```
Comment (isaachinman) [inline src/utils/lng-path-corrector.ts]: Same feedback here for `as.pathname`.

File: src/utils/lng-path-corrector.ts [hunk 6]
```diff
@@ -58,19 +66,37 @@ export default (config: Config, currentRoute, currentLanguage) => {
...
+      as = `/${subpath}${currentAs}`.replace(/\/$/, '')
+    }
+
+    if (typeof href.query !== 'object') throw new Error(`'href.query' must be an object`)
+    if (!href.query) href.query = {}
+
+
     href.query.lng = currentLanguage
-    href.query.subpath = subpath
+    href.query.subpath = subpath || ''
```
Comment (isaachinman) [inline src/utils/lng-path-corrector.ts]: Can you explain why this is necessary?

File: src/utils/lng-path-corrector.ts [hunk 7]
```diff
@@ -58,19 +66,37 @@ export default (config: Config, currentRoute, currentLanguage) => {
...
+    if (typeof as === 'object') {
+      currentAs = as.pathname.replace(basePath, '')
+      as.pathname = `/${subpath}${currentAs}`.replace(/\/$/, '')
+    } else if (typeof as === 'string') {
+      currentAs = as.replace(basePath, '')
+      as = `/${subpath}${currentAs}`.replace(/\/$/, '')
+    }
+
+    if (typeof href.query !== 'object') throw new Error(`'href.query' must be an object`)
+    if (!href.query) href.query = {}
```
Comment (isaachinman) [inline src/utils/lng-path-corrector.ts]: Besides the formatting, which needs to be fixed, this looks pretty dangerous to me. A falsy check isn't good enough - let's be clear about our assertions here.

Also, you're throwing if `href.query` isn't an `object` the line before this. Can you explain the logic here?

**Comment by martpie** [comment]
Sorry for the delay, I will try to address some of your comments and reply to some others asap

> These changes make me a little nervous as they increase complexity and seem to be relying on some unclear/unsafe type assertions.

Regarding the complexity, maybe I can try to simplify things a little bit and be a bit less imperative/sequential.

Regarding the unsafe type assertions, you're right and I will try to improve these. I based myself a lot on TypeScript type checking + Next.js's typings, but there may be runtime issues indeed.

**Comment by isaachinman** [comment]
Hi @martpie - do you want to proceed with this one?

**Comment by martpie** [comment]
By proceeding, do you mean taking it over? or discarding it? 😄

**Comment by isaachinman** [comment]
@martpie Sure, I can take it over if you don't want to continue. Just let me know.

**Comment by martpie** [comment]
Dropping it, it's a bit outdated and I will restart a branch from scratch, addressing some PR comments. I will also check how `next` implemented their link component as there may be some interesting things to get there.

**Pull request closed without merging**